codesearch/server/server.go

```diff
@@ -272,13 +315,24 @@ func (css *codesearchServer) Index(ctx context.Context, req *inpb.IndexRequest)
 
 		log.Infof("Starting indexing %q@%s", repoURL, commitSHA)
 
-		r, err := css.syncIndex(ctx, req)
-		if err != nil {
-			log.Errorf("Failed indexing %q: %s", repoURL, err)
-			return err
+		if req.GetReplacementStrategy() == inpb.ReplacementStrategy_INCREMENTAL {
+			r, err := css.incrementalUpdate(ctx, req)
+			if err != nil {
+				log.Errorf("Failed indexing %q: %s", req.GetGitRepo().GetRepoUrl(), err)
+				return err
+			}
+			rsp = r
+		} else if req.GetReplacementStrategy() == inpb.ReplacementStrategy_REPLACE_REPO {
+			r, err := css.fullyReindex(ctx, req)
+			if err != nil {
+				log.Errorf("Failed indexing %q: %s", req.GetGitRepo().GetRepoUrl(), err)
+				return err
+			}
+			rsp = r
+		} else {
+			return status.InvalidArgumentErrorf("Invalid replacement strategy %s", req.GetReplacementStrategy())
 		}
-		rsp = r
-		log.Infof("Finished indexing %q@%s", repoURL, commitSHA)
+		log.Infof("Finished indexing %s at commit %s", req.GetGitRepo().GetRepoUrl(), req.GetRepoState().GetCommitSha())
 		return nil
 	})
 	if req.GetAsync() {
```

Comment:
Just calling myself out here -- this "spawn a goroutine and potentially never listen for it" shit is extremely corny and I just added it as a quick hack to ensure index requests never blocked. Sorry?

Reply:
I think it's pretty reasonable for a search index to support async updates, especially for something potentially long like a full reindex - is there a better pattern for "keep doing work after the server responds to the initial request"?

Reply:
I guess we could send back a job id or something, and support querying the job status... but def. overkill for where we are

Reply:
I think this is fine as an MVP -- I just hate that nothing is ensuring these goroutines ever finish etc.

(A slightly better solution might be to have a chan owned by the server that we dump stuff in and cap the max number of in-flight updates? Anyway, let's leave it for now, this CL is big enough as is)